src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs

```diff
@@ -111,23 +118,28 @@ private static IEnumerable<IMeasurementGroup> ParseEventData(IEnumerable<IEventM
             var partitionId = data.FirstOrDefault()?.PartitionId;
 
             // Deserialize events into measurements and then group according to the device, type, and other factors
-            return data.Select(e => JsonConvert.DeserializeObject<Measurement>(System.Text.Encoding.Default.GetString(e.Body.ToArray())))
+            return data.Select(e =>
+                {
+                    var measurement = JsonConvert.DeserializeObject<Measurement>(System.Text.Encoding.Default.GetString(e.Body.ToArray()));
+                    measurement.Payload = e.Body;
+                    return measurement;

```

Comment:
It's hard to tell easily, but are we not going to provide the customer with the original message (device payload) when errors occur in the Measurement to FHIR transformation?

Reply:
Right now for FHIR Transformation we will not provide original device message, it is the measurement group payload (though it may be updated to be the individual measurement payload). I would like to provide the original device message since it would be really useful, but the challenge is that I do not think we can guarantee the original device message + normalized message will fit in the normalized event hub given the 1MB limit. Even if we use compression the compression ratio varies depending on the payload content and there could be cases where we would exceed 1MB.

One option could be to just throw an error in the case above and write the original device message to a dead letter queue. One implication though is that we'd be telling users that the message size limit is now (size of input message + size of normalized message) < ~1MB which is less that current state and it may impact users that are close to 1MB already.

Another option could be to not include the original device message if including it would exceed the limit but that could be a inconsistent experience.

The measurement/measurement group should be enough to troubleshoot, but on the other hand there would be nothing to resubmit... so maybe one of the other options is preferable?

Reply:
I think having the original message is very important. The measurement can always be recreated using the codepath that created the measurement in the first place - The customer can do this with our mapping tools, or we can generate the a new measurement after the fact. This is not true for the device message, if we don't include it, it's gone. There are quite a few error scenarios (unexpected values, missing values, etc.) where having the device message would be the only way to figure out what is happening.

Reply:
based on a previous conversation we are going to focus on normalization for now and include the device message for normalization errors. we are going to leave it as a future item on how to best include the device message for measurementtofhir.

Reply:
Are there any user stories/features tracking this future work?